File: clients/packages/api/src/requestFailure.ts
/**
 * One sentence a person can act on, for any request that failed (#263).
 *
 * The API writes its refusals for the user: a JSON `detail` is the whole
 * explanation, and it always wins. Everything else that can go wrong is not
 * the API talking — a proxy in front of it refusing this network, nothing
 * listening at the address, a certificate the app does not trust, a timeout —
 * and axios describes each of those as "Request failed with status code 403"
 * or "Network Error", which names nothing a person can check. Each kind is
 * reduced to what answered, with what code, and what to look at.
 *
 * A wire-protocol refusal (426) keeps `describeMismatch`'s sentence (#150).
 * Synchronous on purpose: the login screen and the settings sections call it
 * from a catch, and the bodies that matter there are already parsed JSON. A
 * Blob body (speech) has its own describer.
 */

import { WIRE_PROTOCOL } from '@kurisu/models';
import { config } from './config';
import { describeMismatch } from './wireProtocol';

interface AxiosLike {
  response?: { status?: number; data?: unknown };
  request?: unknown;
  code?: string;
  message?: string;
}

/** The API's own sentence out of a response body, when there is one. */
export function apiDetail(data: unknown): string | null {
  let body: unknown = data;
  if (typeof body === 'string') {
    try {
      body = JSON.parse(body);
    } catch {
      return null;
    }
  }
  if (body && typeof body === 'object' && 'detail' in body) {
    const detail = (body as { detail?: unknown }).detail;
    if (typeof detail === 'string' && detail.trim()) return detail.trim();
  }
  return null;
}

function isCertificateFailure(err: AxiosLike): boolean {
  const code = (err.code ?? '').toUpperCase();
  const message = (err.message ?? '').toUpperCase();
  return (
    code.startsWith('ERR_CERT_') ||
    code.startsWith('CERT_') ||
    code === 'DEPTH_ZERO_SELF_SIGNED_CERT' ||
    code === 'SELF_SIGNED_CERT_IN_CHAIN' ||
    code === 'UNABLE_TO_VERIFY_LEAF_SIGNATURE' ||
    code === 'ERR_TLS_CERT_ALTNAME_INVALID' ||
    message.includes('ERR_CERT_') ||
    message.includes('SELF SIGNED CERTIFICATE') ||
    message.includes('SELF_SIGNED_CERT') ||
    message.includes('CERTIFICATE HAS EXPIRED')
  );
}

function origin(): string {
  const base = config.apiBaseUrl;
  try {
    return new URL(base).origin;
  } catch {
    return base || 'the server address';
  }
}

/**
 * What to show for `err`. `fallback` is the caller's own wording for the case
 * nothing else fits — the error carried no message at all.
 */
export function describeRequestFailure(err: unknown, fallback = 'Something went wrong.'): string {
  if (!err || typeof err !== 'object') {
    return typeof err === 'string' && err.trim() ? err.trim() : fallback;
  }
  const e = err as AxiosLike;

  if (e.response) {
    const detail = apiDetail(e.response.data);
    if (detail) return detail;

    const status = e.response.status ?? 0;
    switch (status) {
      case 401:
        return 'The server refused the credentials.';
      case 403:
        return "Something in front of the server refused this device (HTTP 403). Check the server address, and whether the operator's proxy allows your network.";
      case 404:
        return 'No KurisuAssistant server answers at this address (HTTP 404).';
      case 426: {
        const body = (e.response.data ?? {}) as { server_wire_protocol?: unknown };
        const serverWire = Number(body.server_wire_protocol);
        return describeMismatch(WIRE_PROTOCOL, Number.isFinite(serverWire) ? serverWire : null);
      }
      case 502:
      case 503:
      case 504:
        return `The server is not reachable behind its proxy (HTTP ${status}).`;
      default:
        if (status >= 500) return `The server failed (HTTP ${status}).`;
        if (status > 0) return `The server answered HTTP ${status}.`;
        return fallback;
    }
  }

  if (isCertificateFailure(e)) {
    return "The server's certificate is not trusted by this app.";
  }
  if (e.code === 'ECONNABORTED' || e.code === 'ETIMEDOUT' || /timeout/i.test(e.message ?? '')) {
    return 'The server did not answer in time.';
  }
  if (e.request || e.code === 'ERR_NETWORK' || e.code === 'ECONNREFUSED' || e.code === 'ENOTFOUND') {
    return `Nothing answered at ${origin()}. Check the address and that the server is running.`;
  }

  const message = e.message?.trim();
  return message ? message : fallback;
}

File: clients/android/app/src/main/java/com/kurisu/assistant/domain/net/RequestFailure.kt
package com.kurisu.assistant.domain.net

import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.security.cert.CertPathValidatorException
import java.security.cert.CertificateException
import javax.net.ssl.SSLHandshakeException
import javax.net.ssl.SSLPeerUnverifiedException
import retrofit2.HttpException

/**
 * One sentence a person can act on, for a request that failed for a reason the
 * API did not write (#263).
 *
 * The API's own `detail` is read by `apiErrorMessage` before this is asked;
 * what reaches here is everything else — a proxy in front of the server
 * refusing this network with its own HTML page, nothing listening at the
 * address, a certificate the app does not trust, a timeout. Retrofit and the
 * JDK describe those as "HTTP 403 Forbidden" or "Connection refused", which
 * names nothing a person can check. The desktop's `requestFailure.ts` says the
 * same sentences.
 */
object RequestFailure {

    /** The sentence for a status the API gave no detail for, or null when the code is not one we name. */
    fun describeStatus(status: Int): String? = when (status) {
        401 -> "The server refused the credentials."
        403 -> "Something in front of the server refused this device (HTTP 403). Check the server address, and whether the operator's proxy allows your network."
        404 -> "No KurisuAssistant server answers at this address (HTTP 404)."
        502, 503, 504 -> "The server is not reachable behind its proxy (HTTP $status)."
        else -> when {
            status >= 500 -> "The server failed (HTTP $status)."
            status > 0 -> "The server answered HTTP $status."
            else -> null
        }
    }

    /**
     * The sentence for a transport failure, or null when `t` is not one — an
     * `HttpException` (the API answered) or an exception of the app's own.
     * `origin` is the server address the user typed, named so they know what
     * to check.
     */
    fun describe(t: Throwable, origin: String?): String? {
        if (t is HttpException) return describeStatus(t.code())
        if (isCertificateFailure(t)) return "The server's certificate is not trusted by this app."
        if (t is SocketTimeoutException) return "The server did not answer in time."
        if (t is ConnectException || t is UnknownHostException) {
            val where = origin?.takeIf { it.isNotBlank() } ?: "the server address"
            return "Nothing answered at $where. Check the address and that the server is running."
        }
        return null
    }

    private fun isCertificateFailure(t: Throwable): Boolean {
        var cause: Throwable? = t
        var hops = 0
        while (cause != null && hops < 8) {
            if (cause is SSLHandshakeException || cause is SSLPeerUnverifiedException ||
                cause is CertPathValidatorException || cause is CertificateException
            ) return true
            cause = cause.cause
            hops++
        }
        return false
    }
}
